Adding Etcd client and limiter

Tobias4872 opened the pull request:
Here is a list of changes that I have done:
- Added redis instance to `docker-compose.yml` to run the tests against.
- Added an Etcd client based on their REST API V3 - there are a few Etcd clients on npmjs.org, but I voted against them as most of them do not seem to be maintained frequently.
- Added a RateLimiter class with the necessary implementation of the given methods.
- Added tests for the new RateLimiter class - I have to admit that I inherited them from the Redis RateLimiter class instead of writing some on my own.

On another note: I have seen the comments about atomic and non-atomic store updates - while Etcd does not have a nice why of updating / increasing values in a single request, I found a way of ensure a transaction-like behavior by using [this](https://etcd.io/docs/v3.4/learning/api/#transaction) approach.

Let me know if you find any issues.

Tobias4872 commented:
Sorry, I missed the interface definitions, but have added them now.

Tobias4872 commented:
And one more note - I occassionally see the "block expires in blockDuration seconds" test failing on both the Memcache and the Redis limiter. Looks like a timing problem / race condition as it does not always fail.

animir commented:
@Tobias4872 Hi, that was quick! 
I left a couple of comments on your PR.
Thanks.

Tobias4872 commented:
Thanks a lot for looking at the PR so quickly. Your comments make total sense for me. I have changed the PR by doing this:

- removed the custom etcd client
- added the etcd3 package which is officially suggestion on the etcd.io package under libraries
- adjusted rate limiter to use the new library
- adjusted tests

Let me know if you find anything else.

animir commented on lib/RateLimiterEtcd.js:
```diff
@@ -0,0 +1,113 @@
+
+      for (let i = 0; i < MAX_TRANSACTION_TRIES; i++) {
+        // eslint-disable-next-line no-await-in-loop
+        oldValue = await this._get(rlKey);
```
@Tobias4872 Could you clarify how it is working in transaction? As I understand it gets data and then puts it. There is some time between get and put when the second get request may be sent to Ectd and it will return the same oldValue as the first one. Is that correct?

Tobias4872 replied:
It first pulls the old value and then sets the new value conditionally - it will only set the new value in case the old value is still present. So, let's assume the key holds the value 1. It will pull that value first and hold it in `oldValue`. And now the magic happens. When putting the new value (let's assume 2), it will first check if the old value 1 is still present. If it is it sets the value 2 and the transaction will succeed. If the value in the store is e.g. 3, the comparison will fail, and the logic starts again up to MAX_TRANSACTION_TRIES. Hope this makes sense - unfortunately, there is no atomic increase or add command in etcd.

Tobias4872 replied:
It is documented [here](https://etcd.io/docs/v3.4/learning/api/).

animir replied:
Interesting. Thank you for the explanation.
It is going to be some kind of bottleneck under the high traffic since only one transaction can succeed during a period of time.

Question: Do you need atomic increments? If not you could rename the class to RateLimiterEtcdNonAtomic and implement simple get/set logic. Let me know what you think about that. We could have both `RateLimiterEtcd` and `RateLimiterEtcdNonAtomic` if that makes sense.

Tobias4872 replied:
Good idea :-) I have implemented that now by adding the suggested `RateLimiterEtcdNonAtomic` and by using that is the base class for `RateLimiterEtcd` which just overrides the `_upsert` method with the "transactional" logic.

animir replied:
Good idea!

Tobias4872 commented on lib/RateLimiterEtcd.js:
```diff
@@ -0,0 +1,113 @@
+      }
+
+      if (!success) {
+        throw new Error('Could not set new value in a transaction.');
```
Makes sense - changed.

Tobias4872 commented on lib/RateLimiterEtcd.js:
```diff
@@ -0,0 +1,114 @@
+      }
+
+      if (!success) {
+        throw new RateLimiterQueueError('Could not set new value in a transaction.');
```
I am sorry, that I didn't catch your intention. I now implemented that error class and also added a `RateLimiterSetupError` which is thrown in case the `storeClient` attribute is not being set. Hope you like that.

animir commented on index.js:
```diff
@@ -36,4 +37,5 @@ module.exports = {
   RateLimiterValkey,
   RateLimiterValkeyGlide,
   RateLimiterSQLite,
+  RateLimiterEtcd,
```
@Tobias4872 `RateLimiterEtcdNonAtomic` should be exported here as well.

animir commented on lib/index.d.ts:
```diff
@@ -513,3 +513,12 @@ export class RateLimiterValkeyGlide extends RateLimiterStoreAbstract {
+ *
+ * The option "storeClient" needs to be set to an instance of class "EtcdClient".
+ */
+export class RateLimiterEtcd extends RateLimiterStoreAbstract {
```
@Tobias4872 `RateLimiterEtcdNonAtomic` Typescript class should be added here as well.

animir replied:
@Tobias4872 Should it be like this
```js
export class RateLimiterEtcd extends RateLimiterEtcdNonAtomic
```
RateLimiterEtcd extends RateLimiterEtcdNonAtomic

animir commented:
@Tobias4872 I think we are close to release—just a couple of small changes required.

Tobias4872 commented:
Sorry, forgot about the index.js / typescript. Added this now.

Any chance you want to add something to the README.md or a separate wiki page?

animir commented:
@Tobias4872 There is one tiny thing in TS and it should be good to go.

animir commented:
@Tobias4872 Thank you for your patience.

animir commented:
@Tobias4872 I added the wiki https://github.com/animir/node-rate-limiter-flexible/wiki/Etcd you prepared and the link in readme. Thanks for the description on how transactions work!

animir commented:
This PR was released in `v7.1.0`

Tobias4872 commented:
Thanks so much and it was rather your patience which made the cooperation a worthwhile experience.
